browser/ui/webui/brave_rewards_ui.cc

```diff
@@ -610,7 +610,22 @@ void RewardsDOMHandler::SaveSetting(const base::ListValue* args) {
     args->GetString(1, &value);
 
     if (key == "enabledMain") {
-      rewards_service_->SetRewardsMainEnabled(value == "true");
+      bool rewards_enabled = (value == "true");
+
+      if (ads_service_) {
+        if (rewards_enabled) {
+          // When Rewards are enabled, set Ads enabled to whatever UI value is
+          bool ads_ui_enabled = ads_service_->is_ui_enabled();
+          ads_service_->set_ads_enabled(ads_ui_enabled);
+        } else {
+          // When Rewards are disabled, cache UI value and disable Ads
+          bool ads_enabled = ads_service_->is_enabled();
```

Comment:
I think it might be better to put this logic in rewards service and AdsServiceFactory should have a DependsOn for RewardsService